ports/esp32/machine_adc.c

```
@@ -182,18 +155,31 @@ void madc_init_helper(const machine_adc_obj_t *self, size_t n_pos_args, const mp
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all(n_pos_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
-    mp_int_t atten = args[ARG_atten].u_int;
-    if (atten != -1) {
-        madc_atten_helper(self, atten);
-    } else if (madc_atten_get(self) == ADC_ATTEN_MAX) {
-        madc_atten_helper(self, ADC_ATTEN_DB_0);
+
+    if (!self->block->handle) {
+        adc_oneshot_unit_init_cfg_t init_config = {
+            .unit_id = self->block->unit_id
+        };
+        check_esp_err(adc_oneshot_new_unit(&init_config, &self->block->handle));
     }
+
+    mp_int_t atten = args[ARG_atten].u_int;
+    mp_machine_adc_atten_set_helper(self, atten != -1 ? atten : ADC_ATTEN_MAX);
+    mp_machine_adc_block_width_set_helper(self->block, ADC_WIDTH_MAX);
+    apply_self_adc_channel_atten(self, mp_machine_adc_atten_get_helper(self));
+
 }
 
 static void mp_machine_adc_init_helper(machine_adc_obj_t *self, size_t n_pos_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
     madc_init_helper(self, n_pos_args, pos_args, kw_args);
 }
 
+static void mp_machine_adc_deinit(machine_adc_obj_t *self) {
+    if (self->block->handle) {
+        check_esp_err(adc_oneshot_del_unit(self->block->handle));
```

If I do `adc.deinit()` twice in a row then the second one fails with an exception.  Ideally it would not, because usually it's expected that `deinit()` (and others like `close()`) don't raise if called multiple times.

Maybe set `self->block->handle = NULL` here after deiniting?

---

Sorry, my mistake. Just added this and tested. Now no error, and adc.read() when deinit'ed returns 0 read values.

---

Are you using some more verbose build settings? I have tested this on the ESP32, S2, S3, C3 using this script:
```py
import machine
a = machine.ADC(0)

print(a.read(),a.read_uv())

a.deinit()
print(a.read(),a.read_uv())

# deinit again on purpose 
a.deinit()
print(a.read(),a.read_uv())

a.init()
print(a.read(),a.read_uv())
```

I do not get the ESP error, just the base reading for read_uv etc. I used the fix you suggested unless I misunderstood :/  
This is the output as per the test script: 
```
2559 2169000
0 128000
0 128000
2549 2145000
```

---

I'm using IDF 5.4.1.

Checking out this PR, rebasing on latest master (rebase is clean), compiling ESP32_GENERIC and running your above test I get:
```
MicroPython v1.26.0-preview.129.g8e8f77718 on 2025-05-20; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> import platform
>>> platform.platform()
'MicroPython-1.26.0-preview-xtensa-IDFv5.4.1-with-newlib4.3.0'
>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== import machine
=== a = machine.ADC(0)
=== 
=== print(a.read(),a.read_uv())
=== 
=== a.deinit()
=== print(a.read(),a.read_uv())
=== 
=== # deinit again on purpose 
=== a.deinit()
=== print(a.read(),a.read_uv())
=== 
=== a.init()
=== print(a.read(),a.read_uv())
13 138000
E (17428) adc_oneshot: adc_oneshot_read(184): invalid argument: null pointer
E (17428) adc_oneshot: adc_oneshot_read(184): invalid argument: null pointer
0 128000
E (17438) adc_oneshot: adc_oneshot_read(184): invalid argument: null pointer
E (17438) adc_oneshot: adc_oneshot_read(184): invalid argument: null pointer
0 128000
176 265000
>>> 
```

---

Instead of printing a message it should raise an exception, eg:
```c
if (!self->block->handle) {
    mp_raise_OSError(MP_EPERM);
}
```